File: 操作系统/Windows API/Windows 程序设计/第6版/CPlusPlus/Chapter06/KeypadWithViewModel/KeypadWithViewModel/DelegateCommand.cpp
#include "pch.h"
#include "DelegateCommand.h"

using namespace Platform;
using namespace Windows::Foundation;

using namespace KeypadWithViewModel;

// Two constructors
DelegateCommand::DelegateCommand(ExecuteDelegate^ execute, CanExecuteDelegate^ canExecute)
{
    executeAction = execute;
    canExecuteAction = canExecute;
}

DelegateCommand::DelegateCommand(ExecuteDelegate^ execute)
{
    executeAction = execute;
    canExecuteAction = ref new CanExecuteDelegate(this, &DelegateCommand::AlwaysCanExecute);
}

// Methods required by ICommand
void DelegateCommand::Execute(Object^ param)
{
    executeAction(param);
}

bool DelegateCommand::CanExecute(Object^ param)
{
    return canExecuteAction(param);
}

// Method required by IDelegateCommand
void DelegateCommand::RaiseCanExecuteChanged()
{
    CanExecuteChanged(this, nullptr);
}

// Default CanExecute method
bool DelegateCommand::AlwaysCanExecute(Object^ param)
{
    return true;
}

File: 操作系统/Windows API/Windows 程序设计/第6版/CSharp/Chapter06/KeypadWithViewModel/KeypadWithViewModel/DelegateCommand.cs
using System;

namespace KeypadWithViewModel
{
    public class DelegateCommand : IDelegateCommand
    {
        Action<object> execute;
        Func<object, bool> canExecute;

        // Event required by ICommand
        public event EventHandler CanExecuteChanged;
        
        // Two constructors
        public DelegateCommand(Action<object> execute, Func<object, bool> canExecute)
        {
            this.execute = execute;
            this.canExecute = canExecute;
        }
        public DelegateCommand(Action<object> execute)
        {
            this.execute = execute;
            this.canExecute = this.AlwaysCanExecute;
        }

        // Methods required by ICommand
        public void Execute(object param)
        {
            execute(param);
        }
        public bool CanExecute(object param)
        {
            return canExecute(param);
        }

        // Method required by IDelegateCommand
        public void RaiseCanExecuteChanged()
        {
            if (CanExecuteChanged != null)
                CanExecuteChanged(this, EventArgs.Empty);
        }

        // Default CanExecute method
        bool AlwaysCanExecute(object param)
        {
            return true;
        }
    }
}
